src/image/channel_groups.rs

```diff
@@ -201,67 +250,74 @@ struct ReadChannelGroups<ReadChannelGroup> {
 }
 
 struct ChannelGroupsReader<ChannelGroupReader> {
-    channels: ChannelGroups<usize>,
-    indexed_channels: Vec<ChannelGroupReader>,
+    channels: ChannelGroups<ChannelGroupReader>,
+
+    // TODO optimize by iterating a vec instead of the nested groups:
+    //channels: Groups<usize>,
+    //indexed_channels: Vec<ChannelGroupReader>,
 }
 
 impl<'s, ReadChannelGroup> ReadChannels<'s> for ReadChannelGroups<ReadChannelGroup>
     where ReadChannelGroup: ReadChannels<'s>
 {
     type Reader = ChannelGroupsReader<ReadChannelGroup::Reader>;
 
-    fn create_channels_reader(&'s self, header: &Header) -> Result<Self::Reader> {
-        let swap = |(a,b)| (b,a);
-        let channel_groups = parse_channel_list_groups(
-            header.channels.list.iter().enumerate().map(swap)
-        );
+    fn create_channels_reader(&'s self, header: &Header, selected_channels_indices: &ChannelMask) -> Result<Self::Reader> {
+        let selected_channels_indices = selected_channels_indices
+            .selected_channel_indices().collect::<SmallVec<[usize; 20]>>();
 
-        let mut indexed_channels = Vec::new();
-        let channel_groups = channel_groups.map(|channels| {
+        // indices refer to `selected_channels_indices`
+        let channel_groups = ChannelGroups::parse_list_to_indices(
+            selected_channels_indices.iter()
+                .map(|&index| &header.channels.list[index])
+                .map(|selected_channel| selected_channel.name.clone())
+        );
 
-            let mut channels_header = header.clone(); // TODO no clone?
-            channels_header.channels = ChannelList::new(channels.iter().map(|(name, index)|{
-                let mut channel_info = header.channels.list[index].clone();
-                channel_info.name = name;
-                channel_info
-            }).collect()); // FIXME does not comply to `header.chunk_count` and that stuff?? change ReadChannels fn signature?
+        Ok(ChannelGroupsReader {
+            // own_channels_indices refer to `selected_channels_indices`
+            channels: channel_groups.try_map(|group_own_channel_indices|{
 
-            indexed_channels.push(self.read_channels.create_channels_reader(&channels_header));
+                let group_selected_channel_indices = group_own_channel_indices.iter()
+                    .map(|&index| selected_channels_indices[index]);
 
-            // FIXME this is not the original order indexed_channels.len() - 1
-            indexed_channels[]
-        });
+                let group_selected_channel_indices = ChannelMask::only(group_selected_channel_indices);
 
-        Ok(ChannelGroupsReader {
-            channels: channel_groups,
-            indexed_channels,
+                let reader = self.read_channels.create_channels_reader(header, &group_selected_channel_indices);
```

Comment:
At this point, meta data contains fully qualified names, but the specific channels and any channels must work with relative local names. Pass some kind of string map instead of only indices?